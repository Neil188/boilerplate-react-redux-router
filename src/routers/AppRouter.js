import React from 'react';
import { Router, Route, Switch } from 'react-router-dom';
import createHistory from 'history/createBrowserHistory';
import DashboardPage from '../components/DashboardPage';
import NotFoundPage from '../components/NotFoundPage';
import ConLoginPage from '../components/LoginPage';
import ConPrivateRoute from './PrivateRoute';
import ConPublicRoute from './PublicRoute';

export const history = createHistory();

const AppRouter = () => (
    <Router history={history} >
        <div>
            <Switch>
                <ConPublicRoute
                    exact
                    path='/'
                    component={ConLoginPage}
                />
                <ConPrivateRoute
                    path='/dashboard'
                    component={DashboardPage}
                />
                <Route
                    component={NotFoundPage}
                />
            </Switch>
        </div>
    </Router>
)

export default AppRouter;
