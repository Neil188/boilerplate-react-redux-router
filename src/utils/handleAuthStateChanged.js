import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { login, logout } from '../actions/auth';
import AppRouter, { history } from '../routers/AppRouter';

const target = document.getElementById('app');

export const jsx = store => (
    <Provider store={store} >
        <AppRouter />
    </Provider>
)

let hasRendered  = false;
export const renderApp = (store) => {
    if (!hasRendered) {
        ReactDOM.render(jsx(store), target);
        hasRendered = true;
    }
}

const handleAuthStateChanged = (store) => user => {
    if (user) {
        store.dispatch( login(user.uid) );
        renderApp(store);
        history.location.pathname === '/' && history.push('/dashboard');
    } else {
        store.dispatch( logout() );
        renderApp(store);
        history.push('/');
    }
}

export default handleAuthStateChanged;
