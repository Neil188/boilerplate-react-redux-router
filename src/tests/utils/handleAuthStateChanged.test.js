import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import handleAuthStateChanged,  { renderApp, LoadableProvider }
    from '../../utils/handleAuthStateChanged';
import { login, logout } from '../../actions/auth';
import { history } from '../../routers/AppRouter';

jest.mock('react-dom');
jest.mock('../../actions/auth');
jest.mock('../../routers/AppRouter');

const store = {
    dispatch: jest.fn(),
    getState: jest.fn(),
    subscribe: jest.fn()
}

test('should render LoadableProvider correctly', () => {
    const wrapper = shallow(<LoadableProvider store={store} />);
    expect(wrapper).toMatchSnapshot();
});

test('on first call renderApp should call ReactDOM', () => {
    renderApp(store);
    expect(ReactDOM.render).toBeCalled();
});

test('if already rendered do not call render again', () => {
    ReactDOM.render.mockClear();
    renderApp(store);
    expect(ReactDOM.render).not.toBeCalled();
});

test('Handle user logged out auth state change', done => {
    handleAuthStateChanged(store)(null)
        .then( () => {
            expect(store.dispatch).toBeCalled();
            expect(logout).toBeCalled();
            expect(login).not.toBeCalled();
            expect(history.push).toHaveBeenLastCalledWith('/');
            done();
        })
});

test('Handle user logged in auth state change, with redirect', done => {
    const user = {
        uid: 1
    }
    logout.mockClear();
    login.mockClear();
    history.location.pathname = '/'

    handleAuthStateChanged(store)(user)
        .then( () => {
            expect(logout).not.toBeCalled();
            expect(login).toBeCalled();
            expect(login).toHaveBeenLastCalledWith(user.uid);
            expect(history.push).toHaveBeenLastCalledWith('/dashboard');
            done();

        })
});

test('Handle user logged in auth state change, without redirect', () => {
    const user = {
        uid: 1
    }
    history.location.pathname = '/test'
    history.push.mockClear();

    handleAuthStateChanged(store)(user);

    expect(history.push).not.toBeCalled();
});
