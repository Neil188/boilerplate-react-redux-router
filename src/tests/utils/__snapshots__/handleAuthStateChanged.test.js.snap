// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render jsx correctly 1`] = `<AppRouter />`;
